Add: Player configuration of badge display.

PeterN opened the pull request:
<!--
Commit message:

- Please use Feature / Add / Change / Fix for player-facing changes. E.g.: "Feature: My cool new feature".
- Please use Feature / Add / Change / Fix followed by "[NewGRF]" or "[Script]" for moddable changes. E.g.: "Feature: [NewGRF] My cool new NewGRF addition".
- Please use Codechange / Codefix for developer-facing changes. E.g.: "Codefix #1234: Validate against nullptr properly".

See https://github.com/OpenTTD/OpenTTD/blob/master/CODINGSTYLE.md#commit-message for more details.
-->

## Motivation / Problem

Whether and where to display badges is designed to be the player's choice. 

Players can prioritise and display the things they are interested in.

<!--
Describe here shortly
* For bug fixes:
    * What problem does this solve?
    * If there is already an issue, link the issue, otherwise describe the problem here.
* For features or gameplay changes:
    * What was the motivation to develop this feature?
    * Does this address any problem with the gameplay or interface?
    * Which group of players do you think would enjoy this feature?
-->


## Description

This PR adds the base configuration system, and adds configuration to the build vehicle window. It can be added to other build windows later.

The button is labelled "B" for now, in the absence of a specific icon for it.

![image](https://github.com/user-attachments/assets/91fddf5a-a902-4ec4-9b71-319fd9f18dbd)

<!--
Describe here shortly
* For bug fixes:
    * How is the problem solved?
* For features or gameplay changes:
    * What does this feature do?
    * How does it improve/solve the situation described under 'motivation'.
-->


## Limitations

<!--
Describe here
* Is the problem solved in all scenarios?
* Is this feature complete? Are there things that could be added in the future?
* Are there things that are intentionally left out?
* Do you know of a bug or corner case that does not work?
-->


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

frosch123 commented on src/dropdown.cpp:
```diff
@@ -87,6 +87,9 @@ struct DropdownWindow : Window {
 
 	Dimension items_dim{}; ///< Calculated cropped and padded dimension for the items widget.
 
+	static inline Dimension selected_item_dimension;
```
I would prefer if these were not stored globally:
* `DropDownListItem`/`DropDownMover` should get a `OnClick` method, which returns a `button_index`.
* This keeps the pixel position stuff in a single class, instead of splitting it between `DropDownMover` and `HandleBadgeConfigurationDropDownClick`.
* The button can then be passed to `OnDropdownSelect` as third parameter.

frosch123 commented on src/newgrf_badge_config.cpp:
```diff
@@ -0,0 +1,201 @@
+		"badges_industries", // GSF_INDUSTRIES
+		{}, // GSF_CARGOES
+		{}, // GSF_SOUNDFX
+		"badge_columns_airports", // GSF_AIRPORTS
```
This one does not match the rest:
```suggestion
		"badges_airports", // GSF_AIRPORTS
```

frosch123 commented on src/newgrf_badge_config.cpp:
```diff
@@ -0,0 +1,201 @@
+			if (badge.name == STR_NULL) continue;
+			if (!badge.features.Test(feature)) continue;
+
+			/* Search for this label in the vehicle badge configuration. */
```
```suggestion
```

frosch123 commented on src/newgrf_badge_config.cpp:
```diff
@@ -0,0 +1,201 @@
+			if (found != std::end(config)) continue;
+
+			/* Not found, insert it. */
+			config.emplace_back(badge.label, 0, 0, true);
```
`container_func.hpp` has a `include` function.
Is it useful to add an overload with an predicate?

PeterN replied:
Dunno, I considered `include()` to be a left over helper from when we didn't use vectors. It doesn't fit our naming convention either...

frosch123 commented on src/newgrf_badge_config.cpp:
```diff
@@ -0,0 +1,201 @@
+	if (found == std::end(config)) return empty;
+
+	/* Sort order is simply the position in the configuration list. */
+	found->sort_order = static_cast<int>(std::distance(std::begin(config), found));
```
Shoudn't this be in `AddBadgeClassesToConfiguration`?

frosch123 commented on src/newgrf_badge_config.h:
```diff
@@ -0,0 +1,32 @@
+	std::string label; ///< Class label.
+	int column = -1; ///< UI column, feature-dependent.
+	int sort_order = -1; ///< Position within list, populated only by GetBadgeClassConfigItem.
+	bool show_icon = false; ///< Set if the badge icons should be displayed for this class.
```
In `GUIBadgeClasses::Element` this is called `visible`.

PeterN replied:
`show_icon` because `show_filter` will also be a thing.

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -76,12 +79,10 @@ GUIBadgeClasses::GUIBadgeClasses(GrfSpecFeature feature)
 
-		this->gui_classes.emplace_back(class_index, column, visible, sort_order, size, class_badge->label);
-		if (visible) max_column = std::max<uint>(max_column, column);
+		this->gui_classes.emplace_back(class_index, config.column, config.show_icon, config.sort_order, size, class_badge->label);
```
Can `config.column` be negative here? It's initialised with `-1`.

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+
+static constexpr int CLICK_MOVE_UP = 1;
+static constexpr int CLICK_MOVE_DOWN = 2;
+static constexpr int CLICK_TOGGLE_ICON = 3;
```
```suggestion
static constexpr int CLICK_TOGGLE_VISIBLE = 3;
```

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+
+using DropDownListToggleMoverItem = DropDownMover<DropDownToggle<DropDownString<DropDownListItem>>>;
+
+static constexpr int CLICK_MOVE_UP = 1;
```
An `enum` would auto-increment consecutive members.

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+		}
+
+		Rect br = r.WithWidth(w, TEnd ^ rtl).CentreTo(w, SETTING_BUTTON_HEIGHT);
+		DrawArrowButtons(br.left, br.top, this->button_colour, state, this->click_up, this->click_down);
```
```suggestion
		DrawArrowButtons(br.left, br.top, this->button_colour, state, this->click_up != 0, this->click_down != 0);
```
Not sure whether the "clickable" state needs swapping for RTL.

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+	auto it = std::ranges::find(config, class_badge.label, &BadgeClassConfigItem::label);
+	if (it == std::end(config)) return;
+
+	if (click_result == CLICK_TOGGLE_ICON) it->show_icon = !it->show_icon;
```
Why is this check here, and not in the `switch` in `HandleBadgeConfigurationDropDownClick`?

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+{
+	DropDownList list;
+
+	list.push_back(MakeDropDownListStringItem(STR_BADGE_CONFIG_RESET, INT_MAX));
```
Maybe add a `static constexpr int ITEM_RESET = INT_MAX;`?

frosch123 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -245,3 +246,205 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+{
+	if (result == INT_MAX) {
+		ResetBadgeClassConfiguration(feature);
+		return false;
```
Doesn't resetting also need a widget reinit? I.e. `return true`?

PeterN committed:
Codechange: Add OnClick handler for dropdown items.

This allows each dropdown item to indicate if something different should happen depending on where in the item was clicked.

PeterN committed:
Codechange: Add Up/Down buttons, to fit in with the settings-button style.

PeterN committed:
Codechange: Add DropDownToggle component.

PeterN committed:
Add: Generic "configure list/view" vertical ellipsis icon.

PeterN committed:
Codechange: Pass click state to Dropdown Draw() function.

This allows drawing to change depending on last click position.

PeterN committed:
Add: Button to configure badges in build vehicle window.

PeterN committed:
Add: Configuration of NewGRF badges.

PeterN committed:
Add: Button to configure badges in build picker window.

rubidium42 commented on src/textfile_gui.h:
```diff
@@ -31,9 +31,9 @@ struct TextfileWindow : public Window, MissingGlyphSearcher {
 	void OnInit() override;
 	void OnInvalidateData(int data = 0, bool gui_scope = true) override;
-	void OnDropdownSelect(WidgetID widget, int index) override;
-	void OnRealtimeTick(uint delta_ms) override;
```
Why is this moved? I reckon something went awry during resolving a conflict.

rubidium42 commented on src/dropdown_common_type.h:
```diff
@@ -182,6 +183,56 @@ class DropDownCheck : public TBase {
+		bool rtl = TEnd ^ (_current_text_dir == TD_RTL);
+		int w = SETTING_BUTTON_WIDTH;
+
+		if (r.WithWidth(w, rtl).CentreTo(w, SETTING_BUTTON_HEIGHT).Contains(pt)) return this->click;
```
`CentreTo` is added two commits later than this, so this and the next commit do not compile.

rubidium42 commented on src/newgrf_badge_gui.cpp:
```diff
@@ -246,3 +247,210 @@ std::unique_ptr<DropDownListItem> MakeDropDownListBadgeIconItem(const std::share
+	if (it->column < pos_next->column) {
+		++it->column;
+	} else {
+		/* Rotate elements left so that it is placed after itnext, maintaining order of non-visible elements. */
```
What's the `itnext` supposed to be?

PeterN committed:
Codechange: Add CentreTo Rect method.
